thewalrus/_hafnian.py

```diff
@@ -178,6 +180,45 @@ def hafnian(
     )
 
 
+def hafnian_sparse(A, D = None):
+    r"""Returns the hafnian of a sparse symmetric matrix.
+    This pure python implementation is very slow on full matrices, but faster the sparser a matrix is.
+    As a rule of thumb, the crossover in runtime with respect to hafnian() happens around 50% sparsity.
+
+    Args:
+        A (array): the symmetric matrix of which we want to compute the hafnian
+        D (set): set of indices that identify a submatrix. If None (default) it computes
+        the hafnian of the whole matrix.
+
+    Returns:
+        (float) hafnian of A or of the submatrix of A defined by the set of indices D.
+    """
+    if D is None:
+        D = frozenset(range(len(A)))
+    else:
+        D = frozenset(D)
+
+    if np.allclose(A, np.zeros_like(A)):
+        return 0.0
+
+    r, _ = np.nonzero(A)
+    m = max(Counter(r).values())  # max nonzero values per row/column
+
+    @lru_cache(maxsize=2 ** m)
+    def indices(d, k):
+        return d.intersection(set(np.nonzero(A[k])[0]))
+
+    @lru_cache(maxsize=2 ** m)
+    def lhaf(d: frozenset) -> float:
+        if not d:
+            return 1
+        d_ = set(d)
+        k = d_.pop()
+        return sum(A[i, k] * lhaf(frozenset(d_).difference({i})) for i in indices(d, k))
```

Comment:
While brief, maybe this line would be more readable using a for loop and a list.

Reply:
I understand your point, but this line reads like the mathematical formula. I think I prefer it this way.